core/src/main/java/org/hisp/dhis/android/core/program/internal/ProgramIndicatorCall.kt

```diff
@@ -0,0 +1,91 @@
+ *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.android.core.program.internal
+
+import dagger.Reusable
+import io.reactivex.Single
+import javax.inject.Inject
+import org.hisp.dhis.android.core.arch.api.executors.internal.APIDownloader
+import org.hisp.dhis.android.core.arch.call.factories.internal.UidsCall
+import org.hisp.dhis.android.core.arch.handlers.internal.Handler
+import org.hisp.dhis.android.core.common.ObjectWithUid
+import org.hisp.dhis.android.core.program.ProgramIndicator
+
+@Reusable
+internal class ProgramIndicatorCall @Inject constructor(
+    private val service: ProgramIndicatorService,
+    private val handler: Handler<ProgramIndicator>,
+    private val apiDownloader: APIDownloader,
+    private val programStore: ProgramStoreInterface
+) : UidsCall<ProgramIndicator> {
+
+    companion object {
+        const val MAX_UID_LIST_SIZE = 50
+    }
+
+    override fun download(uids: Set<String>): Single<List<ProgramIndicator>> {
+        val programUids = programStore.selectUids()
+
+        val firstPayload = apiDownloader.downloadPartitioned(
+            uids = programUids.toSet(),
+            pageSize = MAX_UID_LIST_SIZE,
+            pageDownloader = { partitionUids ->
+                val displayInFormFilter = ProgramIndicatorFields.displayInForm.eq(true)
+                val programUidsFilter = "program.${ObjectWithUid.uid.`in`(partitionUids).generateString()}"
+                service.getProgramIndicator(
+                    fields = ProgramIndicatorFields.allFields,
+                    displayInForm = displayInFormFilter,
+                    program = programUidsFilter,
+                    uids = null,
+                    false
+                )
+            }
+        )
+
+        val secondPayload = apiDownloader.downloadPartitioned(
+            uids = uids,
+            pageSize = MAX_UID_LIST_SIZE,
+            pageDownloader = { partitionUids ->
+                service.getProgramIndicator(
+                    fields = ProgramIndicatorFields.allFields,
+                    displayInForm = null,
+                    program = null,
+                    uids = ProgramIndicatorFields.uid.`in`(partitionUids),
+                    false
+                )
+            }
+        )
+
+        return Single.merge(firstPayload, secondPayload).reduce { t1, t2 ->
```

Comment:
The handling of the items is done within the `reduce` body. It will work in this case because there are only two singles, but if we add a third one, like `Single.merge(firstPayload, secondPayload, thirdPayload)`, the code will compile and will look ok but the handling will be done multiple 2 times:
- First time (t1 + t2)
- Second time ((t1 + t2) + t3)

It will be better to use other operator to wait for all the calls to complete. Or reduce the items and do the handling outside the reduce function.